language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -219,23 +208,38 @@ dummyType = TyVar () dummyTyName
 
 -- This works using type environments to handle substitutions efficiently. We
 -- keep a type environment which holds all substitutions which should be in
--- scope at any given momeny. After any lookups, we clone the looked-up type in
+-- scope at any given moment. After any lookups, we clone the looked-up type in
 -- order to maintain global uniqueness.
 --
 -- This is all tracked in a state monad, and we simply delete any substitutions
 -- once they go out of scope; this is permissible since 'Unique's are globally
 -- unique and so we will not delete the wrong thing.
 -- | Synthesize the type of a term, returning a normalized type.
 typeOf :: Term TyNameWithKind NameWithType a -> TypeCheckM a (NormalizedType TyNameWithKind ())
-typeOf (Var _ (NameWithType (Name (_, ty) _ _))) = normalizeTypeOpt $ void ty
-typeOf (LamAbs _ _ ty t)                         =
-    TyFun () <<$>> normalizeTypeOpt (void ty) <<*>> typeOf t
+typeOf (Var _ (NameWithType (Name (_, ty) _ _))) =
+    -- Since we kind check types at lambdas, we can normalize types here without kind checking.
+    -- Type normalization at each variable is inefficient and we may consider something else later.
+    normalizeTypeOpt $ void ty
+typeOf (LamAbs x _ dom body)                     = do
+    -- [check| dom :: *]    dom ~>? vDom    [infer| body : vCod]
+    -- ---------------------------------------------------------
+    -- [infer| lam n dom body : vDom -> vCod]
+
+    kindCheckM x dom $ Type ()
```

Comment:
The type inference rules for `(lam y T M)` do not currently require that `T` have kind `(type)`. Is this something we discussed? The spec should be updated if this is the case.

Reply:
The rule in the spec is

```
Γ, y : T |- M : S
--------------------
Γ |- (lam y T M ) : (fun T S)
```

`fun T S` is simply ill-kinded if `T` is not of kind `*`. But maybe the spec should indeed say

> In context Γ, term M has normal well-kinded type S

instead of

> In context Γ, term M has normal type S

Just for clarity. @psygnisfive, what do you think?

Reply:
@effectfully Yes I agree that this check should be part of the hypotheses (and hence, is appropriate in the code).

Reply:
Then I'll add a bit more docs and merge. Thanks.

Reply:
Yes, I agree that it is best to update the spec and implementation in tandem.

Reply:
While I think it's nice to keep the spec in sync with the implementation, that would also slow down the development process considerably. Changing the spec (and any theoretical work in general) is slow by its nature. And we want development to be fast.

Regarding this particular issue: it's at worst a documentation problem in the spec. No updates to theory are required. And the plutus-metatheory repo handles everything correctly. It is also obviously correct that the kind of the type of the variable a lambda binds should be `*`, because only types of this kind are inhabited by terms. So I do not want this PR to be blocked on this.

Reply:
To further clarify, I think, it has been a flaw in our process that the spec changes have been so slow going. One goal of having the spec alongside the code in the same repo is to change this.

If you are unsure of how to change the LaTeX, please let me know and I'll add the diff.

Reply:
I'm very reluctant to change the spec myself. I'll add those few symbols to the spec, but
1) I'm not confident they're needed there
2) there could be a better description in a different place

Reply:
Done. Also this introduces a merge conflict with the recent PR by @psygnisfive.

Reply:
> And merge conflicts are a fact of life if we want to make good progress as a team of our current size.

Sure, I'm just CC'ed Rebecca, so she knows where the conflict comes from.

> I was actually thinking of also extending the inference rule to

I added kind-checking to `Error`, `TyInst` and `Wrap` as well, so these cases should be changed too. I'll start a conversation about this in slack.

Reply:
Ahhhh interesting. I'm going to add the following judgment-defining rules to the spec:

`--------------
Γ ⊢ α :: K jdg`

`Γ ⊢ A :: type`
-------------
Γ ⊢ x : A jdg`

Does this make sense to everyone? Or should I make it slightly less formal b/c of the risk of weird circularity?

Reply:
Is there a formal approach that avoids circularity?